File: backend/internal/services/route_service.go
package services

import (
	"database/sql"
	"fmt"
	"sts-backend/internal/database"
	"sts-backend/internal/models"
)

// RouteService handles business logic for route management
type RouteService struct {
	repo *database.RouteRepository
}

func NewRouteService(db *sql.DB) *RouteService {
	return &RouteService{repo: database.NewRouteRepository(db)}
}

func (s *RouteService) GetAllRoutes() ([]models.MasterRoute, error) {
	return s.repo.GetAll()
}

func (s *RouteService) GetRouteStops(routeID string) ([]models.RouteStop, error) {
	return s.repo.GetRouteStops(routeID)
}

func (s *RouteService) GetRouteByID(id string) (*models.MasterRoute, error) {
	route, err := s.repo.GetByID(id)
	if err != nil {
		return nil, err
	}
	if route == nil {
		return nil, fmt.Errorf("route not found")
	}
	return route, nil
}

func (s *RouteService) CreateRoute(req *models.CreateRouteRequest) (*models.MasterRoute, error) {
	if req.EncodedPolyline == "" {
		return nil, fmt.Errorf("encoded_polyline is required")
	}
	if req.RouteName == "" {
		return nil, fmt.Errorf("route_name is required")
	}
	return s.repo.Create(req)
}

func (s *RouteService) UpdateRoute(id string, req *models.UpdateRouteRequest) (*models.MasterRoute, error) {
	route, err := s.repo.Update(id, req)
	if err != nil {
		return nil, err
	}
	if route == nil {
		return nil, fmt.Errorf("route not found")
	}
	return route, nil
}

func (s *RouteService) DeleteRoute(id string) error {
	return s.repo.Delete(id)
}

File: frontend/src/app/core/services/route.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { MasterRoute, CreateRouteRequest, UpdateRouteRequest } from '../models/route.model';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class RouteService {
  private baseUrl = `${environment.apiUrl}/routes`;

  constructor(private http: HttpClient) {}

  getAllRoutes(): Observable<MasterRoute[]> {
    return this.http.get<MasterRoute[]>(this.baseUrl);
  }

  getRouteById(id: string): Observable<MasterRoute> {
    return this.http.get<MasterRoute>(`${this.baseUrl}/${id}`);
  }

  getRouteStops(id: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/${id}/stops`);
  }

  createRoute(req: CreateRouteRequest): Observable<MasterRoute> {
    return this.http.post<MasterRoute>(this.baseUrl, req);
  }

  updateRoute(id: string, req: UpdateRouteRequest): Observable<MasterRoute> {
    return this.http.put<MasterRoute>(`${this.baseUrl}/${id}`, req);
  }

  deleteRoute(id: string): Observable<any> {
    return this.http.delete<any>(`${this.baseUrl}/${id}`);
  }
}
